### PR TITLE
Modify address view model

### PR DESCRIPTION
**Pull request by neelSharma12**
This PR contains refactored code for AddressViewmodel to accommodate  the customized UI of billing address for various components. I am attaching the customized UI screenshot as well as default billing address UI.

### CODE REVIEW

File: AdyenCard/Components/Card/CardComponent.swift [hunk 1]
```diff
@@ -49,6 +49,9 @@ public class CardComponent: PublicKeyConsumer,
     /// Card component configuration.
     public let configuration: Configuration
     
+    /// Card component address builder.
+    public let addressViewModelBuilder: AddressViewModelBuilder
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardComponent.swift]: maybe this should be `internal` or `private`

File: AdyenDropIn/DropInComponent.swift [hunk 2]
```diff
@@ -49,12 +52,15 @@ public final class DropInComponent: NSObject,
     ///   - configuration: The payment method specific configuration.
     ///   - title: Name of the application. To be displayed on a first payment page.
     ///            If no external value provided, the Main Bundle's name would be used.
+    ///   - addressViewModelBuilder: The  address viewmodel builder
     public init(paymentMethods: PaymentMethods,
                 configuration: Configuration,
-                title: String? = nil) {
+                title: String? = nil,
+                addressViewModelBuilder: AddressViewModelBuilder) {
```
Comment (mohammedDehairy) [inline AdyenDropIn/DropInComponent.swift]: this shouldn't be injected by merchants

File: AdyenCard/Components/Card/CardViewControllerItemsProvider.swift [hunk 3]
```diff
@@ -34,7 +36,8 @@ extension CardViewController {
                       cardLogos: [FormCardLogosItem.CardTypeLogo],
                       scope: String,
                       defaultCountryCode: String,
-                      localizationParameters: LocalizationParameters?) {
+                      localizationParameters: LocalizationParameters?,
+                      addressViewModelBuilder: AddressViewModelBuilder) {
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardViewControllerItemsProvider.swift:40]: card component should just use the default `AddressViewModelBuilder` without being injected

File: AdyenDropIn/Utilities/ComponentManager.swift [hunk 4]
```diff
@@ -167,7 +172,8 @@ internal final class ComponentManager {
 
         return BCMCComponent(paymentMethod: paymentMethod,
                              apiContext: apiContext,
-                             configuration: configuration)
+                             configuration: configuration,
+                             addressViewModelBuilder: addressViewModelBuilder)
```
Comment (mohammedDehairy) [inline AdyenDropIn/Utilities/ComponentManager.swift]: all components that needs the `addressViewModelBuilder` should decide which implementation they want in its internal logic and not receive it from outside, merchants shouldn't care about this to inject it into components :-)

File: AdyenCard/Components/Card/CardComponent.swift (hunk 1 above)
Comment (neelSharma12) [reply AdyenCard/Components/Card/CardComponent.swift]: yes, you are right I missed this. I have updated the code. Thanks for pointing it out.

File: AdyenDropIn/DropInComponent.swift (hunk 2 above)
Comment (neelSharma12) [reply AdyenDropIn/DropInComponent.swift]: yes, you are right. I have updated the code. Thanks for pointing it out.

File: AdyenCard/Components/Card/CardViewControllerItemsProvider.swift (hunk 3 above)
Comment (neelSharma12) [reply AdyenCard/Components/Card/CardViewControllerItemsProvider.swift:40]: Yes. I have updated the code.

File: AdyenDropIn/Utilities/ComponentManager.swift (hunk 4 above)
Comment (neelSharma12) [reply AdyenDropIn/Utilities/ComponentManager.swift]: I missed this point. Thanks a lot for letting me know.

File: Adyen/UI/Form/Items/Address/AtomeAddressViewModelBuilder.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
+//
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+public struct AtomeAddressViewModelBuilder: AddressViewModelBuilder {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/AtomeAddressViewModelBuilder.swift]: maybe this should be in the Adyen module and make it internal

File: Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift [hunk 6]
```diff
@@ -0,0 +1,71 @@
+//
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+public protocol AddressViewModelBuilder {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift:10]: maybe add /// :nodoc:

File: Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    func build(countryCode: String) -> AddressViewModel
+}
+
+public struct DefaultAddressViewModelBuilder: AddressViewModelBuilder {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift:15]: maybe add /// :nodoc:

File: Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    
+    public init() {}
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift:18]: maybe add /// :nodoc:

File: Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    
+    // swiftlint:disable function_body_length explicit_acl
+    public func build(countryCode: String) -> AddressViewModel {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/DefaultAddressViewModelBuilder.swift:22]: maybe add /// :nodoc:

**Review by mohammedDehairy — DISMISSED** [review]

File: Adyen/UI/Form/Items/Address/AddressViewModel.swift [hunk 10]
```diff
@@ -17,72 +18,30 @@ internal enum AddressField: String, CaseIterable {
     case country
 }
 
-internal enum FormScheme {
+/// :nodoc:
+public enum FormScheme {
```
Comment (descorp) [inline Adyen/UI/Form/Items/Address/AddressViewModel.swift]: Maybe `AddressFormScheme` ?

Comment (neelSharma12) [reply Adyen/UI/Form/Items/Address/AddressViewModel.swift]: I have renamed this. thanks for pointing it out.

**Review by descorp — DISMISSED** [review]

**Review by nauaros — DISMISSED** [review]
Nice one!

**Review by descorp — APPROVED** [review]

**Review by nauaros — APPROVED** [review]

**Pull request merged**